RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/SiPixelRawToClusterKernel.dev.cc

```diff
@@ -0,0 +1,810 @@
+        ALPAKA_ASSERT_OFFLOAD(clus_view[1024].moduleStart() >= clus_view[1023].moduleStart());
+        ALPAKA_ASSERT_OFFLOAD(clus_view[1025].moduleStart() >= clus_view[1024].moduleStart());
+        ALPAKA_ASSERT_OFFLOAD(clus_view[nMaxModules].moduleStart() >= clus_view[1025].moduleStart());
+
+        cms::alpakatools::for_each_element_in_block_strided(acc, nMaxModules + 1, [&](uint32_t i) {
+          if (0 != i)
+            ALPAKA_ASSERT_OFFLOAD(clus_view[i].moduleStart() >= clus_view[i - i].moduleStart());
+          // Check BPX2 (1), FP1 (4)
+          constexpr auto bpix2 = TrackerTraits::layerStart[1];
+          constexpr auto fpix1 = TrackerTraits::layerStart[4];
+          if (i == bpix2 || i == fpix1)
+            printf("moduleStart %d %d\n", i, clus_view[i].moduleStart());
+        });
+#endif
+        // avoid overflow
+        constexpr auto MAX_HITS = TrackerTraits::maxNumberOfHits;
+        cms::alpakatools::for_each_element_in_block_strided(acc, nMaxModules + 1, [&](uint32_t i) {
+          if (clus_view[i].clusModuleStart() > MAX_HITS)
+            clus_view[i].clusModuleStart() = MAX_HITS;
+        });
+
+      }  // end of fillHitsModuleStart kernel operator()
+    };   // end of fillHitsModuleStart struct
+
+    // Interface to outside
+    template <typename TrackerTraits>
+    void SiPixelRawToClusterKernel<TrackerTraits>::makePhase1ClustersAsync(
+        const SiPixelClusterThresholds clusterThresholds,
+        const SiPixelMappingSoAConstView &cablingMap,
+        const unsigned char *modToUnp,
+        const SiPixelGainCalibrationForHLTSoAConstView &gains,
+        const WordFedAppender &wordFed,
+        const uint32_t wordCounter,
+        const uint32_t fedCounter,
+        bool useQualityInfo,
+        bool includeErrors,
+        bool debug,
+        Queue &queue) {
+      nDigis = wordCounter;
+
+#ifdef GPU_DEBUG
+      std::cout << "decoding " << wordCounter << " digis." << std::endl;
+#endif
+      constexpr int numberOfModules = TrackerTraits::numberOfModules;
+      digis_d = SiPixelDigisSoACollection(wordCounter, queue);
+      if (includeErrors) {
+        digiErrors_d = SiPixelDigiErrorsSoACollection(wordCounter, queue);
+      }
+      clusters_d = SiPixelClustersSoACollection(numberOfModules, queue);
+      if (wordCounter)  // protect in case of empty event....
+      {
+        const int threadsPerBlockOrElementsPerThread = []() {
+          if constexpr (std::is_same_v<Device, alpaka_common::DevHost>) {
+            // NB: MPORTANT: This could be tuned to benefit from innermost loop.
+            return 32;
+          } else {
+            return 512;
+          }
+        }();
+        // fill it all
+        const uint32_t blocks = cms::alpakatools::divide_up_by(wordCounter, threadsPerBlockOrElementsPerThread);
+        const auto workDiv = cms::alpakatools::make_workdiv<Acc1D>(blocks, threadsPerBlockOrElementsPerThread);
+        ALPAKA_ASSERT_OFFLOAD(0 == wordCounter % 2);
```

Comment:
Since this is host code, maybe this should be
```suggestion
        assert(0 == wordCounter % 2);
```